File: backend/app/schemas/schedule.py
"""日程模块 Pydantic Schemas"""
from __future__ import annotations

from datetime import date, datetime
from pydantic import BaseModel, Field


# ─── Schedule Event ────────────────────────────────────────────────────────────

class ScheduleEventCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=100)
    event_type: str = Field(..., pattern="^(game|training|internal|other)$")
    start_date: date
    end_date: date
    description: str | None = Field(default=None, max_length=2000)

    def model_post_init(self, __context: object) -> None:
        if self.end_date < self.start_date:
            raise ValueError("end_date 不能早于 start_date")


class ScheduleEventUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=100)
    event_type: str | None = Field(default=None, pattern="^(game|training|internal|other)$")
    start_date: date | None = None
    end_date: date | None = None
    description: str | None = None


class ScheduleEventRead(BaseModel):
    id: int
    team_id: int
    title: str
    event_type: str
    start_date: date
    end_date: date
    description: str | None
    status: str
    created_by: int
    linked_match_id: int | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ScheduleEventListItem(BaseModel):
    """日历中每个日程的简要信息（用于日历格点展示）"""
    id: int
    title: str
    event_type: str
    start_date: date
    end_date: date
    status: str
    linked_match_id: int | None
    attendance_count: int = 0    # 已提交出勤数
    total_players: int = 0       # 队伍总人数
    yes_count: int = 0
    sdl_count: int = 0
    leave_count: int = 0
    no_count: int = 0
    not_submitted_count: int = 0

    model_config = {"from_attributes": True}


# ─── Attendance ────────────────────────────────────────────────────────────────

class AttendanceSubmit(BaseModel):
    status: str = Field(..., pattern="^(yes|leave|sdl)$")


class AttendanceRead(BaseModel):
    id: int
    event_id: int
    player_id: int
    player_name: str
    player_display_name: str | None
    status: str
    submitted_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class AttendanceSummary(BaseModel):
    """管理员查看出勤汇总"""
    event_id: int
    yes: list[AttendanceRead] = []
    leave: list[AttendanceRead] = []
    sdl: list[AttendanceRead] = []
    not_submitted: list[dict] = []  # {"player_id": int, "player_name": str}


# ─── Line Division ─────────────────────────────────────────────────────────────

class LinePlayerInfo(BaseModel):
    player_id: int
    player_name: str
    display_name: str | None
    conservative_rating: float
    gender: str | None = None
    jersey_number: int | None = None

    model_config = {"from_attributes": True}


class ScheduleLineRead(BaseModel):
    id: int
    line_name: str
    line_type: str
    round_number: int
    order_index: int
    players: list[LinePlayerInfo] = []

    model_config = {"from_attributes": True}


class ScheduleLineDivisionRead(BaseModel):
    id: int
    event_id: int
    division_method: str
    total_rounds: int
    lines: list[ScheduleLineRead] = []

    model_config = {"from_attributes": True}


class DivisionCreate(BaseModel):
    """创建/重置分 line 方案"""
    division_method: str = Field(default="manual", pattern="^(manual|auto_balanced|auto_strong_to_weak)$")
    total_rounds: int = Field(default=1, ge=1, le=10)


class DivisionUpdate(BaseModel):
    """更新分 line 方案轮数（仅支持增加）"""
    total_rounds: int = Field(..., ge=1, le=10)


class LineCreate(BaseModel):
    """在方案下新建一条 line"""
    line_name: str = Field(..., min_length=1, max_length=50)
    line_type: str = Field(default="line", pattern="^(o_line|d_line|line)$")
    round_number: int = Field(default=1, ge=1)
    order_index: int = Field(default=0, ge=0)


class LineUpdate(BaseModel):
    line_name: str | None = Field(default=None, min_length=1, max_length=50)
    line_type: str | None = Field(default=None, pattern="^(o_line|d_line|line)$")
    order_index: int | None = None


class LinePlayerAdd(BaseModel):
    player_id: int


class AutoAssignRequest(BaseModel):
    """自动分line请求"""
    method: str = Field(..., pattern="^(auto_balanced|auto_strong_to_weak)$")
    num_lines: int = Field(..., ge=1, le=8)
    round_number: int = Field(default=1, ge=1)
    # 仅包含这些球员（可以是 yes + sdl + leave），不传则取出勤为 yes 的
    player_ids: list[int] | None = None


class ScheduleLineTemplateSave(BaseModel):
    template_name: str = Field(..., min_length=1, max_length=50)


class ScheduleLineTemplateRead(BaseModel):
    id: int
    event_type: str
    template_name: str
    line_count: int = 0
    updated_at: datetime

    model_config = {"from_attributes": True}


class SmartLineAnalyzeRequest(BaseModel):
    player_ids: list[int] = []
    schedule_event_id: int | None = None
    apply_to_match: bool = False
    max_line_size: int = Field(default=7, ge=3, le=20)
    d_line_count: int = Field(default=1, ge=1, le=2)
    handler_ratio: int = Field(default=3, ge=1, le=5)
    cutter_ratio: int = Field(default=4, ge=1, le=6)
    recent_matches: int = Field(default=6, ge=3, le=20)


class ManualLineInput(BaseModel):
    line_name: str
    line_type: str  # "o_line" | "d_line" | "line"
    player_ids: list[int]


class ManualLineAnalyzeRequest(BaseModel):
    lines: list[ManualLineInput]
    recent_matches: int = Field(default=6, ge=3, le=20)
    handler_ratio: int = Field(default=3, ge=1, le=5)
    cutter_ratio: int = Field(default=4, ge=1, le=6)


class SmartLinePlayerRead(BaseModel):
    player_id: int
    player_name: str
    display_name: str | None
    gender: str | None = None
    role: str
    ability_score: float
    chemistry_score: float
    offense_score: float
    scoring_score: float
    recent_form_score: float
    total_score: float
    reason: str


class SmartLineChemistryPairRead(BaseModel):
    player_a_id: int
    player_b_id: int
    player_a_name: str
    player_b_name: str
    chemistry_score: float
    combo_count: int
    co_matches: int
    summary: str


class SmartLineGroupRead(BaseModel):
    line_name: str
    line_type: str
    total_score: float
    chemistry_average: float = 0.0
    player_ids: list[int]
    players: list[SmartLinePlayerRead]
    chemistry_pairs: list[SmartLineChemistryPairRead] = []


class SmartLineAnalyzeResponse(BaseModel):
    event_id: int | None = None
    applied_to_match: bool = False
    lines: list[SmartLineGroupRead] = []
    o_line: SmartLineGroupRead
    d_lines: list[SmartLineGroupRead]
    rationale: dict


# ─── Used by MatchInputView to fetch schedule info for linking ──────────────────

class ScheduleForMatchLink(BaseModel):
    """新建比赛时可选关联的日程信息（精简）"""
    id: int
    title: str
    event_type: str
    start_date: date
    rounds: list[int] = []                   # 内战：可用轮次列表
    lines_by_round: dict[int, list[ScheduleLineRead]] = {}  # 内战：每轮 line 列表

    model_config = {"from_attributes": True}

File: frontend/src/api/schedule.ts
import api from '@/api'

export interface ScheduleEvent {
  id: number
  title: string
  event_type: 'game' | 'training' | 'internal' | 'other'
  status: 'draft' | 'published'
  start_date: string
  end_date: string
  description?: string
  notes?: string
  linked_match_id?: number
  attendance_count?: number
  total_players?: number
  yes_count?: number
  sdl_count?: number
  leave_count?: number
  no_count?: number
  not_submitted_count?: number
}

export interface ScheduleReminderResponse {
  message: string
  reminded: number
  events?: number
}

export interface AttendanceSummary {
  yes: PlayerBrief[]
  leave: PlayerBrief[]
  sdl: PlayerBrief[]
  not_submitted: PlayerBrief[]
  no?: PlayerBrief[] // legacy compatibility only
}

export interface PlayerBrief {
  player_id: number
  player_name: string
  display_name?: string
}

export interface LinePlayerInfo {
  player_id: number
  player_name: string
  display_name?: string
  conservative_rating: number
  gender?: string | null
  jersey_number?: number | null
  attendance_status?: string
}


export interface ScheduleLineRead {
  id: number
  line_name: string
  line_type: string
  round_number: number
  order_index: number
  players: LinePlayerInfo[]
}

export interface ScheduleLineDivisionRead {
  id: number
  event_id: number
  division_method: string
  total_rounds: number
  lines: ScheduleLineRead[]
}

export interface ScheduleLineTemplate {
  id: number
  event_type: 'game' | 'training' | 'internal' | 'other'
  template_name: string
  line_count: number
  updated_at: string
}

export interface DivisionForMatch {
  event_id: number
  event_type: string
  total_rounds: number
  rounds: Record<string, { id: number; line_name: string; line_type: string; player_ids: number[] }[]>
}

export interface SmartLinePlayer {
  player_id: number
  player_name: string
  display_name?: string | null
  gender?: string | null
  role: 'handler' | 'cutter'
  ability_score: number
  chemistry_score: number
  offense_score: number
  scoring_score: number
  recent_form_score: number
  total_score: number
  reason: string
}

export interface SmartLineChemistryPair {
  player_a_id: number
  player_b_id: number
  player_a_name: string
  player_b_name: string
  chemistry_score: number
  combo_count: number
  co_matches: number
  summary: string
}

export interface SmartLineGroup {
  line_name: string
  line_type: string
  total_score: number
  chemistry_average: number
  player_ids: number[]
  players: SmartLinePlayer[]
  chemistry_pairs: SmartLineChemistryPair[]
}

export interface SmartLineAnalyzeResponse {
  event_id?: number | null
  applied_to_match: boolean
  lines: SmartLineGroup[]
  o_line: SmartLineGroup
  d_lines: SmartLineGroup[]
  rationale: Record<string, unknown>
}

// ─── Events ───────────────────────────────────────────────────────────────────

export const scheduleApi = {
  getEvents: (params: { start_date?: string; end_date?: string; event_type?: string }) =>
    api.get<ScheduleEvent[]>('/schedule-events', { params }).then(r => r.data),

  getEvent: (id: number) =>
    api.get<ScheduleEvent>(`/schedule-events/${id}`).then(r => r.data),

  createEvent: (body: Partial<ScheduleEvent>) =>
    api.post<ScheduleEvent>('/schedule-events', body).then(r => r.data),

  updateEvent: (id: number, body: Partial<ScheduleEvent>) =>
    api.put<ScheduleEvent>(`/schedule-events/${id}`, body).then(r => r.data),

  deleteEvent: (id: number) =>
    api.delete(`/schedule-events/${id}`),

  publishEvent: (id: number) =>
    api.post(`/schedule-events/${id}/publish`),

  unpublishEvent: (id: number) =>
    api.post(`/schedule-events/${id}/unpublish`),

  remindEvent: (id: number) =>
    api.post<ScheduleReminderResponse>(`/schedule-events/${id}/remind`).then(r => r.data),

  remindPendingEvents: () =>
    api.post<ScheduleReminderResponse>('/schedule-events/remind/pending').then(r => r.data),

  getLinkableEvents: (matchType?: 'external' | 'internal') =>
    api.get<ScheduleEvent[]>('/schedule-events/for-match/linkable', {
      params: matchType ? { match_type: matchType } : undefined,
    }).then(r => r.data),

  // ─── Attendance ─────────────────────────────────────────────────────────────

  getMyAttendance: (eventId: number) =>
    api.get<{ status: string }>(`/schedule-attendance/${eventId}/me`).then(r => r.data),

  submitMyAttendance: (eventId: number, status: string) =>
    api.put(`/schedule-attendance/${eventId}/me`, { status }),

  getAttendanceSummary: (eventId: number) =>
    api.get<AttendanceSummary>(`/schedule-attendance/${eventId}/summary`).then(r => r.data),

  // ─── Lines ───────────────────────────────────────────────────────────────────

  getDivision: (eventId: number) =>
    api.get<ScheduleLineDivisionRead>(`/schedule-lines/${eventId}/division`).then(r => r.data),

  createOrResetDivision: (eventId: number, body: { division_method: string; total_rounds: number }) =>
    api.post<ScheduleLineDivisionRead>(`/schedule-lines/${eventId}/division`, body).then(r => r.data),

  updateDivision: (eventId: number, body: { total_rounds: number }) =>
    api.put<ScheduleLineDivisionRead>(`/schedule-lines/${eventId}/division`, body).then(r => r.data),

  deleteRound: (eventId: number, roundNumber: number) =>
    api.delete<ScheduleLineDivisionRead>(`/schedule-lines/${eventId}/division/rounds/${roundNumber}`).then(r => r.data),

  createLine: (eventId: number, body: { line_name: string; line_type: string; round_number: number; order_index: number }) =>
    api.post<ScheduleLineRead>(`/schedule-lines/${eventId}/division/lines`, body).then(r => r.data),

  updateLine: (eventId: number, lineId: number, body: { line_name?: string; order_index?: number }) =>
    api.put<ScheduleLineRead>(`/schedule-lines/${eventId}/division/lines/${lineId}`, body).then(r => r.data),

  deleteLine: (eventId: number, lineId: number) =>
    api.delete(`/schedule-lines/${eventId}/division/lines/${lineId}`),

  addPlayerToLine: (eventId: number, lineId: number, playerId: number) =>
    api.post<ScheduleLineRead>(`/schedule-lines/${eventId}/division/lines/${lineId}/players`, { player_id: playerId }).then(r => r.data),

  removePlayerFromLine: (eventId: number, lineId: number, playerId: number) =>
    api.delete(`/schedule-lines/${eventId}/division/lines/${lineId}/players/${playerId}`),

  autoAssign: (eventId: number, body: { method: string; num_lines: number; round_number: number; player_ids?: number[] }) =>
    api.post<ScheduleLineDivisionRead>(`/schedule-lines/${eventId}/division/auto-assign`, body).then(r => r.data),

  listDivisionTemplates: (eventType: 'game' | 'training') =>
    api.get<ScheduleLineTemplate[]>('/schedule-lines/templates', { params: { event_type: eventType } }).then(r => r.data),

  saveDivisionTemplate: (eventId: number, templateName: string) =>
    api.post<ScheduleLineTemplate>(`/schedule-lines/${eventId}/division/templates`, { template_name: templateName }).then(r => r.data),

  applyDivisionTemplate: (eventId: number, templateId: number) =>
    api.post<ScheduleLineDivisionRead>(`/schedule-lines/${eventId}/division/templates/${templateId}/apply`).then(r => r.data),

  getDivisionForMatch: (eventId: number) =>
    api.get<DivisionForMatch>(`/schedule-lines/${eventId}/division/for-match`).then(r => r.data),

  analyzeSmartExternalLines: (
    body: {
      player_ids: number[]
      schedule_event_id?: number
      apply_to_match?: boolean
      max_line_size?: number
      d_line_count?: 1 | 2
      recent_matches?: number
      handler_ratio?: number
      cutter_ratio?: number
    },
  ) => api.post<SmartLineAnalyzeResponse>('/schedule-lines/smart-external-lines', body).then(r => r.data),

  smartExternalLines: (
    eventId: number,
    body: { player_ids: number[]; apply_to_match?: boolean; recent_matches?: number; handler_ratio?: number; cutter_ratio?: number },
  ) => api.post<SmartLineAnalyzeResponse>(`/schedule-lines/${eventId}/division/smart-external-lines`, body).then(r => r.data),

  analyzeAssignedLines: (body: {
    lines: Array<{ line_name: string; line_type: string; player_ids: number[] }>
    recent_matches?: number
    handler_ratio?: number
    cutter_ratio?: number
  }) => api.post<SmartLineAnalyzeResponse>('/schedule-lines/analyze-assigned-lines', body).then(r => r.data),
}

export interface ChemistryPairItem {
  player_a_id: number
  player_b_id: number
  chemistry_score: number
  co_matches: number
}

export async function fetchChemistryPairs(playerIds: number[]): Promise<ChemistryPairItem[]> {
  if (playerIds.filter(id => id > 0).length < 2) return []
  const res = await api.post<ChemistryPairItem[]>('/players/chemistry-pairs', {
    player_ids: playerIds.filter(id => id > 0), // 过滤掉负数 guest id
  })
  return res.data
}

export default scheduleApi
